### ISSUE TITLE
Feature: group_by_streaming (no_std version of group_by)

### ISSUE THREAD

**Issue by jhwgh1968** [rust-itertools/itertools#356]
The current documentation of ``group_by`` states:

> If the groups are consumed in order, or if each group's iterator is dropped without keeping it around, then GroupBy uses no allocations. It needs allocations only if several group iterators are alive at the same time.

If there were a way to guarantee a specific piece of code would never hit those cases, it could be made ``#[no_std]``. To achieve this, I propose a new method: ``group_by_streaming`` (name open for bikeshedding, of course).

It is a slightly less ergonomic version of ``group_by`` which is designed to remain compatible with most cases while preventing allocations from happening.

Basically:

1. The ``group_by_streaming`` method will consume the iterator to guarantee exclusive access.
2. The ``next`` method will return a ``StreamingGroup<'g>`` whose lifetime is bound to the ``GroupByStreaming`` that returned it, guaranteeing only one group exists at a time.
3. The implementation will be pretty much a copy-paste of ``group_by``, except there is no ``Vec<T>`` and code paths that require allocation all become ``panic!`` -- because they should be impossible. 

If this API is accepted, I would gladly implement it.

EDIT: fixed small logic flaw

**Comment by bluss** [comment]
The iterator trait does not allow point 2, and re point 1, the way group by does it today is exactly so that the group's can share access to the base iter using references instead of shared ownership boxes.

Feel free to make a proof of concept, something with more ergonomy instead and more bang (panic!) outside its typical use sounds interesting.

**Comment by jhwgh1968** [comment]
Thanks for the input, @bluss!

I tried to do the obvious thing first: replace the ``Vec<T>`` with a generic ``Backing<T>`` trait, and implement it for both ``Vec`` and a ZST which panics instead of allocating. But that turned out to be a bigger mess that I thought, and pretty unmaintainable. I would basically have to clone the entire Vec API, and anyone refactoring the code would have to update it with new ``Vec`` methods they used.

But if you're okay with "more bang", then I'll start with something else entirely: the implementation I have for [a different proposal with a similar lifetime problem](https://github.com/rust-itertools/itertools/issues/284) that I never released. (Hopefully you can see the similarity in data sharing, even though it preferred to return ``None`` instead of ``panic!``.)

It is pretty ugly, and currently a soundness hole (because ``Pin`` didn't exist yet). But I think I can turn that into something solid.

Hope to have a WIP PR soon!

**Comment by jhwgh1968** [comment]
Sorry I meant to update this a while ago:

Looking back at my other version, I used some heap tricks to prevent a number of iterator invalidation cases. Not only would this not work with most ``#![no_std]`` environments, but I found a couple of other UB holes when I poked at it.

Closing this issue as "cannot be implemented in a memory safe way."

**Comment by bluss** [comment]
A PoC of "streaming" group by could look like this for no_std:  https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=89168810e3fd6c65d1392ff0ef2a4f52

It's essentially the same as the current group_by except it doesn't implement buffering. It enforces that one group is finished before starting the next, by panicking, which makes certain iterator patterns impossible.

This is not an endorsement of the pattern and I wouldn't guarantee that such a hack could be included in itertools :smile:
